feat(aws_instance): add instance market options block

haidargit opened the pull request:
## what

- The PR enables the terraform [instance_market_options](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/instance#instance_market_options) block for this ec2 module  
- add a **new tfvars** as a consumer/_Terratest_ for the ec2 spot instance inside `examples/complete` directory

## why

- **Spot Instances** can offer a great discount compared to **On-Demand** pricing based on AWS documentation

## references

- there are no issues related to this PR, this PR is made for the ec2 module improvement  

others:
- module documentation is updated (guide: [Updating Module Documentation](https://github.com/cloudposse/.github/blob/0544cc9a6cb769f9cb713fa4c26a30e52ec63a6e/CONTRIBUTING.md#updating-module-documentation)) 

- update the `versions.tf` files in the example directories as a complement (guide: [Updating the Module Examples](https://github.com/cloudposse/.github/blob/0544cc9a6cb769f9cb713fa4c26a30e52ec63a6e/CONTRIBUTING.md#updating-the-module-examples)) 

- (**done**) terraform testing (_plan+apply_) is performed by using `examples/complete` directory on current PR branch `feature/spot-instance-enablement` targeting self-owned AWS cloud account

joe-niland commented on variables.tf:
```diff
@@ -46,6 +46,32 @@ variable "burstable_mode" {
+
+variable "market_type" {
+  type        = string
+  description = "Type of market for the instance. Valid values are `spot` and `capacity-block`. Defaults to `spot`. Required if `spot_options` is specified."
```
```suggestion
  description = "Type of market for the instance. Valid values are `spot` and `capacity-block`. Defaults to `spot`. Required if a non-empty value is provided for `spot_options_attributes`."
```

haidargit replied:
suggestion committed, thanks for the input @joe-niland

joe-niland reviewed (commented):
Looks good @haidargit - a couple of comments

haidargit commented on variables.tf:
```diff
@@ -46,6 +46,32 @@ variable "burstable_mode" {
+    Describes the market (purchasing) option for the instances.
+    To get more info see https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/instance#spot-options .
+  EOT
+  default     = null
```
sure, will update this line

haidargit replied:
code updated [[ref](https://github.com/cloudposse/terraform-aws-ec2-instance/pull/202/files#diff-05b5a57c136b6ff596500bcbfdcff145ef6cddea2a0e86d184d9daa9a65a288eR72)]

nitrocode replied:
cc: @haidargit you should be able to remove `instance_market_options_enabled` entirely

haidargit replied:
noted, code updated @nitrocode, @joe-niland

haidargit replied:
docs/make init+make readme updated 👍🏻

haidargit commented:
`make init && make readme` have been re-ran, @joe-niland, Thank you

nitrocode commented:
That new example directory isn't getting run by terratest. Was adding the example strictly as a way of documenting the feature?

Here are some options worth considering

1. How come we can't make the existing tfvars test use the new input? This would be ideal.

2. If that's not possible, is it possible to add a new tfvars in the existing examples complete directory instead of adding a new terraform example directory? If so that would be advisable because it would be less code to maintain and update. 

3. As-is (example directory, no automated test)

---

Here is the existing tfvars https://github.com/cloudposse/terraform-aws-ec2-instance/blob/main/examples/complete/fixtures.us-east-2.tfvars which could then have an additional tfvars beside it such as fixtures.us-east-2.spot.tfvars.

Here is how the tfvars get inserted into the test https://github.com/cloudposse/terraform-aws-ec2-instance/blob/c4be48095dbccb409a776cd5c08959467696c776/test/src/examples_complete_test.go#L27

haidargit commented:
Hi @nitrocode, thanks for the insight

Correct, using the existing example directories is sufficient. 

i have updated my code to use the current `/complete` test directory, these are the changes that may be applicable: 
- i have created the _fixtures.us-east-2.spot.tfvars_ as new tvfars. I just left the current tfvars as is inside the `examples/complete` directory, so the user/terratest able to provision a sample of an EC2 instance with the spot instance disabled. The new _tvfars_ will introduce this new variable `instance_market_options_enabled = true`

- modified the `test/src/examples_complete_test.go` to include new tfvars and spot instance request ID output test

for the terratest reference, the output ID prefix is `sir-`
![spot_instance_request id_sir](https://github.com/user-attachments/assets/3c856584-75ea-4aea-91e8-263f5ca68902)

nitrocode commented on outputs.tf:
```diff
@@ -100,3 +100,13 @@ output "security_group_name" {
+
+output "spot_instance_request_id" {
+  value       = var.instance_market_options_enabled && var.market_type == "spot" ? aws_instance.default[*].spot_instance_request_id : null
+  description = "the ID of the Spot Instance request"
```
```suggestion
  description = "ID of the Spot Instance request"
```

nitrocode commented on variables.tf:
```diff
@@ -46,6 +46,32 @@ variable "burstable_mode" {
 
+variable "instance_market_options_enabled" {
+  type        = bool
+  description = "Wheter to enable the purchasing option for the instances"
```
```suggestion
  description = "Whether to enable the purchasing option for the instances"
```

nitrocode commented:
Thanks for addressing the changes. 

How come the _enabled flag is still used? I thought it was removed when this comment was raised?

https://github.com/cloudposse/terraform-aws-ec2-instance/pull/202#discussion_r1719397421

haidargit commented:
@nitrocode  is it `instance_market_options` for the naming?

nitrocode commented:
/terratest

nitrocode commented:
@haidargit see code annotations and test failures

nitrocode commented:
Thanks for resolving some of the issues.

There seem to be a few remaining. Please fix all of them and then resubmit the PR. I moved it to a draft for now.

haidargit commented:
Hi @nitrocode,
i adjusted the _output.tf_ files on the spot instance part to use _length_ function to fix the tflint warnings

Could you kindly review [the latest commit](https://github.com/cloudposse/terraform-aws-ec2-instance/pull/202/commits/534f46b78db2e0c2ff17e98d003a31f8fc3f9702) when you get a chance? Much appreciated!

haidargit commented:
@nitrocode,
this was spot instance testing conducted by `terraform apply` using `fixtures.us-east-2.spot.tfvars` on `/complete` directory. The PR has also set the default to use normal EC2 instance provisioning (commit 0c3e3b3)

**case 1**:  
spot instance enabled (`market_type = "spot"`) and spot options declared (non empty `spot_options_attributes`)

![1_apply_with_spot](https://github.com/user-attachments/assets/7e76de53-b0ed-4290-a07c-501d5756e4c3)
![Screen Shot 2024-09-22 at 10 52 30](https://github.com/user-attachments/assets/6ed86898-4ece-4ab3-a58d-c5c70e20587a)

<br/>**case 2**:  
spot instance disabled (`market_type = null`).  
As long as the spot instance is disabled (`market_type = null`), the `spot_options_attributes` variable will be ignored regardless of whether its declared or not.

nitrocode commented on variables.tf:
```diff
@@ -46,6 +46,26 @@ variable "burstable_mode" {
 
+variable "market_type" {
+  type        = string
+  description = "(Optional) Type of market for the instance. Valid values are `spot` and `capacity-block`. If the value set is `spot`, the `spot_options_attributes` variable becomes adjustable."
```
The `terraform-docs` automatically marks inputs as `optional` or `required` based on if there is a `default` set.

```suggestion
  description = "Type of market for the instance. Valid values are `spot` and `capacity-block`. If the value set is `spot`, the `spot_options_attributes` variable becomes adjustable."
```

nitrocode commented on variables.tf:
```diff
@@ -46,6 +46,26 @@ variable "burstable_mode" {
+    valid_until                    = string
+  }))
+  description = <<-EOT
+    Describes the market (purchasing) option for the instances.
+    To get more info see https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/instance#spot-options .
```
You can use markdown here to shorten the URL.

```suggestion
    Describes the market (purchasing) option for the instances.
    See [docs](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/instance#spot-options) for more information.
```

nitrocode commented on test/src/examples_complete_test.go:
```diff
@@ -24,7 +24,7 @@ func TestExamplesComplete(t *testing.T) {
 		Upgrade:      true,
 		// Variables to pass to our Terraform code using -var-file options
-		VarFiles: []string{"fixtures.us-east-2.tfvars"},
+		VarFiles: []string{"fixtures.us-east-2.tfvars","fixtures.us-east-2.spot.tfvars"},
```
Instead of expanding on VarFiles, it would be best to create a separate unit test otherwise one of these inputs can stomp over the other inputs.

nitrocode replied:
cc: @cloudposse/engineering for thoughts here

haidargit replied:
does this commit [c34444c](https://github.com/cloudposse/terraform-aws-ec2-instance/pull/202/commits/c34444c037cd6e3151055a5cfbda78ac05298326) apply to this condition?

a new function  `TestSpotInstanceComplete` is added to test the EC2 spot instance

nitrocode commented on outputs.tf:
```diff
@@ -100,3 +100,13 @@ output "security_group_name" {
 }
+
+output "instance_lifecycle" {
+  value       = length(var.spot_options_attributes[*]) != 0 ? aws_instance.default[*].instance_lifecycle : null
```
```suggestion
  value       = try(one(aws_instance.default[*].instance_lifecycle), null)
```

nitrocode commented on outputs.tf:
```diff
@@ -100,3 +100,13 @@ output "security_group_name" {
+}
+
+output "spot_instance_request_id" {
+  value       = length(var.spot_options_attributes[*]) != 0 && var.market_type == "spot" ? aws_instance.default[*].spot_instance_request_id : null
```
```suggestion
  value       = try(one(aws_instance.default[*].spot_instance_request_id), null)
```

nitrocode commented on variables.tf:
```diff
@@ -46,6 +46,26 @@ variable "burstable_mode" {
 
+variable "market_type" {
+  type        = string
+  description = "(Optional) Type of market for the instance. Valid values are `spot` and `capacity-block`. If the value set is `spot`, the `spot_options_attributes` variable becomes adjustable."
```
You may consider custom validation here https://developer.hashicorp.com/terraform/language/values/variables#custom-validation-rules to ensure only `spot` or `capacity-block` is used

haidargit replied:
will do, thanks @nitrocode

haidargit replied:
updated https://github.com/cloudposse/terraform-aws-ec2-instance/pull/202/commits/c18468768c919ad5a9b75338b3e9cd0e7b7dadeb

nitrocode commented on variables.tf:
```diff
@@ -46,6 +46,26 @@ variable "burstable_mode" {
   default     = null
 }
 
+variable "market_type" {
```
Take a look at how the auto scaling group implements both of these inputs 

https://github.com/cloudposse/terraform-aws-ec2-autoscale-group?tab=readme-ov-file#input_instance_market_options

https://github.com/cloudposse/terraform-aws-ec2-autoscale-group/blob/906f7975d4a5eaae9b40b268c988e3aa43e18e3e/variables.tf#L100-L115

haidargit replied:
looks way more clean. 
will adjust the inputs, thanks!

haidargit replied:
updated https://github.com/cloudposse/terraform-aws-ec2-instance/pull/202/commits/c18468768c919ad5a9b75338b3e9cd0e7b7dadeb

nitrocode commented:
I'll cc: @joe-niland here as I need help in reviewing the PR

haidargit commented:
Hello @nitrocode, @joe-niland 

I've updated the code. Could you help review the latest commit? [c184687](https://github.com/cloudposse/terraform-aws-ec2-instance/pull/202/commits/c18468768c919ad5a9b75338b3e9cd0e7b7dadeb)

the changes included the instance market options variable adjustment, custom validation for market type, and several other minor fix

Thank you 🙏🏻

joe-niland commented:
/terratest

nitrocode commented on examples/complete/variables.tf:
```diff
@@ -55,4 +55,25 @@ variable "metric_treat_missing_data" {
+  EOT
+  default     = null
+  validation {
+    condition     = contains(["spot", "capacity-block"], var.instance_market_options.market_type)
```
It's failing on this condition

```hcl
TestExternalEniComplete 2024-10-15T22:07:13Z logger.go:66: │ Error: Attempt to get attribute from null value
TestExternalEniComplete 2024-10-15T22:07:13Z logger.go:66: │ 
TestExternalEniComplete 2024-10-15T22:07:13Z logger.go:66: │   on ../../variables.tf line 65, in variable "instance_market_options":
TestExternalEniComplete 2024-10-15T22:07:13Z logger.go:66: │   65:     condition     = contains(["spot", "capacity-block"], var.instance_market_options.market_type)
TestExternalEniComplete 2024-10-15T22:07:13Z logger.go:66: │     ├────────────────
TestExternalEniComplete 2024-10-15T22:07:13Z logger.go:66: │     │ var.instance_market_options is null
TestExternalEniComplete 2024-10-15T22:07:13Z logger.go:66: │ 
TestExternalEniComplete 2024-10-15T22:07:13Z logger.go:66: │ This value is null, so it does not have any attributes.
TestExternalEniComplete 2024-10-15T22:07:13Z logger.go:66: ╵
TestExternalEniComplete 2024-10-15T22:07:13Z retry.go:99: Returning due to fatal error: FatalError{Underlying: error while running command: exit status 1; ╷
│ Error: Attempt to get attribute from null value
│ 
│   on ../../variables.tf line 65, in variable "instance_market_options":
│   65:     condition     = contains(["spot", "capacity-block"], var.instance_market_options.market_type)
│     ├────────────────
│     │ var.instance_market_options is null
│ 
│ This value is null, so it does not have any attributes.
╵}
    destroy.go:11: 
        	Error Trace:	destroy.go:11
        	            				panic.go:523
        	            				testing.go:999
        	            				apply.go:15
        	            				examples_complete_test.go:109
        	Error:      	Received unexpected error:
        	            	FatalError{Underlying: error while running command: exit status 1; ╷
        	            	│ Error: Attempt to get attribute from null value
        	            	│ 
        	            	│   on ../../variables.tf line 65, in variable "instance_market_options":
        	            	│   65:     condition     = contains(["spot", "capacity-block"], var.instance_market_options.market_type)
        	            	│     ├────────────────
        	            	│     │ var.instance_market_options is null
        	            	│ 
        	            	│ This value is null, so it does not have any attributes.
        	            	╵}
        	Test:       	TestExternalEniComplete
```

nitrocode commented:
/terratest

joe-niland commented:
/terratest
